### PR TITLE
Fix LMMS plugin issues

### LINKED ISSUE #6669: automation max value changes when loading a saved file
### Bug Summary
when i make an automation track with points that have values higher than 1, they will be reverted to 1 upon loading

#### Steps to reproduce
open organic synth
add automation track
drag harmonic dial to automation track (values are 0-17)
lay points on the track varying from 0-17
save file
load file
check automation track. upper limit will be set back to 1.0, and points with values greater than 1 become 1.

#### Expected behavior
automation points should stay at the number they were assigned. not truncated down to 1.0

#### Actual behavior
automation points greater than 1.0, become 1.0

#### Affected LMMS versions
1.2.2

#### Logs
<details>
  <summary>Click to expand</summary>
<pre>

</pre>
</details>

### LINKED ISSUE #6671: Vibed instrument data loss upon saving and reloading project file
### Bug Summary
The Vibed instrument only saves the parameters and sample data for each of its 9 "Strings" to the project file if that particular string is active at the time the project file is saved. So if I edit one of the strings of my Vibed instrument but have it disabled at the time that I save the project (whether it's temporarily disabled, such as in an automation, or not), all that data will be lost. And when I reload the project, the automation track for such a Vibed string's parameters is messed up.

#### Steps to reproduce
- Create new project
- Add Vibed instrument, and open its GUI
- Select String 2, which is not enabled by default
- Draw a shape in the sample area and change some parameters for the string so that it is different from the default, but keep the string disabled
- Save the project and reload from disk
- All changes made to String 2 of the Vibed instrument will be gone

#### Expected behavior
All Vibed instrument data should be preserved regardless of whether it is currently used

#### Actual behavior
Vibed instrument data for String #N is lost if String #N is disabled at the time the project is saved

#### Affected LMMS versions
master branch

#### Logs
<details>
  <summary>Click to expand</summary>
<pre>

</pre>
</details>

### LINKED ISSUE #6797: [Kicker plugin] Automation connection of a particular knob is not restored when reopening a project
Hi,

I automated multiple knobs of a Kicker plugin, using some automation tracks in the song. I noticed that the connection of the "End distortion" is systematically not restored after closing/reopening the project. This never happen with another automated knob of this plugin, but with "End distortion" it happen every times.

Version used : LMMS 1.3.0-alpha.1.315+g57e7fdc53 (Linux x86_64, Qt 5.15.8, GCC 12.2.0)

### PR DESCRIPTION
**Pull request by messmerd**
This PR fixes two related bugs affecting LMMS instrument plugins, refactors the Vibed instrument, and fixes memory leaks.

Fixes #6669
Fixes #6671
Fixes #6797 

## 1) Automated parameter loading bug (#6669, #6797)

The automation for certain parameters of LMMS plugins was failing to load correctly from the project files.

This bug affected the following parameters when they were automated:
- **Organic**'s harmonics parameters
- **Kicker**'s end distortion parameter
- **AudioFileProcessor**'s interpolation parameter
- **Vibed**'s volume parameters

In each of these cases, conditional parameter loading logic was used to check for the existence of a parameter with the given name in the XML data. However, only the existence of the XML attribute was checked, which is how non-automated parameters are stored. To do it correctly, they should have also checked for the existence of an XML child element, which is how automated parameters are stored. This PR fixes that oversight, which resolves the bugs.

Note: A parameter may also be stored as an attribute of a child element called "automatablemodel", but that isn't the case for any of the parameters affected by this bug.

I also took the liberty of cleaning up the style/formatting of the `loadSettings` and `saveSettings` methods.

## 2) Vibed data saving/loading bug (#6671)

The Vibed instrument was only saving the parameters and sample data for each of its 9 "Strings" to the project file if that particular string was active at the time the project file was saved. This led to a couple relatively serious issues including data loss.

I fixed the problem by always saving each of the 9 Vibed string parameters and their sample data regardless of whether they are currently active. I incremented the "version" attribute which Vibed XML data uses so that this change is backwards compatible with older project files.

## 3) Vibed refactor and improvements

The Vibed instrument's pre-C++11 codebase was in great need of an overhaul and since I was already making a number of changes to it, I decided to refactor the whole thing. Besides making the code much more readable, this refactor brings a few additional improvements/fixes such as:
- Fixes a few memory leaks
- Replaces many raw pointers with smart pointers
- Prioritizes stack memory over heap memory for better real-time safety
- Removes some usages of Qt

## 4) Memory leaks
Fixes memory leaks affecting nearly all LMMS instrument plugins during heavy tempo automation. ([See explanation](https://github.com/LMMS/lmms/pull/6670#discussion_r1141261115))

### CODE REVIEW

**Review by Veratil — COMMENTED** [review]
General lookover review, nothing in depth.

File: plugins/Vibed/Vibed.cpp [hunk 1]
```diff
@@ -255,425 +215,326 @@ void Vibed::loadSettings( const QDomElement & _this )
...
 }
 
-
-
-
-void Vibed::playNote( NotePlayHandle * _n, sampleFrame * _working_buffer )
+void Vibed::playNote(NotePlayHandle* n, sampleFrame* workingBuffer)
 {
-	if ( _n->totalFramesPlayed() == 0 || _n->m_pluginData == nullptr )
+	if (/*n->totalFramesPlayed() == 0 ||*/ n->m_pluginData == nullptr)
```
[the remark this answers is not in the document]
Comment (messmerd) [reply plugins/Vibed/Vibed.cpp:228]: All the other instrument plugins use the same condition before creating the object to store in `n->m_pluginData`, but it seems to me like a subtle bug waiting to happen if `n->totalFramesPlayed()` is ever zero while `n->m_pluginData` is non-null, you know? Seems like we should only be checking one not both, but I wasn't completely sure about it which is why I just commented it out.

[the remark this answers is not in the document]
Comment (PhysSong) [reply plugins/Vibed/Vibed.cpp]: > it seems to me like a subtle bug waiting to happen if `n->totalFramesPlayed()` is ever zero while `n->m_pluginData` is non-null

If I remember correctly, that might happen when the tempo is heavily automated.

[the remark this answers is not in the document]
Comment (messmerd) [reply plugins/Vibed/Vibed.cpp]: I just did a more thorough investigation into this.

**TLDR**: Yes, a heavily automated tempo breaks it and causes a memory leak. This memory leak affects almost all instrument plugins.

**Long version**:
In most instrument plugins' `playNote` routines, if the condition `m_totalFramesPlayed == 0 && m_pluginData != nullptr` ever occurs, the `m_pluginData` pointer would be overwritten with a pointer to new dynamically allocated memory and a memory leak would occur. For said condition to occur, `m_instrumentTrack->playNote(...)` would need to have been called previously (which is the only way for `m_pluginData != nullptr` to happen), and `m_totalFramesPlayed` would need to have been set to zero sometime after that. So the question becomes "Is it possible for `m_totalFramesPlayed` be set to zero prior to calling `m_instrumentTrack->playNote(...)` again?" (The answer is yes)

These are the only three places I can find where the value of `m_totalFramesPlayed` changes:
**(1)** The NotePlayHandle constructor initializes it to zero.
**(2)** `NotePlayHandle::play(...)` increases it by `framesThisPeriod` at the end before returning.
**(3)** `NotePlayHandle::resize(...)` sets it to zero when called with `m_origFrames == 0` or `m_origTempo == 0`.

Possibility (1) can be ignored since the constructor is only called once at the start before `m_instrumentTrack->playNote(...)` is ever called, and it sets `m_pluginData` to nullptr, so it cannot cause the condition `m_totalFramesPlayed == 0 && m_pluginData != nullptr`.

As for possibility (2), it looks like `framesThisPeriod` is always >0 when set, so `m_totalFramesPlayed` should always increase at the end of `NotePlayHandle::play(...)`. And if `m_instrumentTrack->playNote(...)` is called, it will create the condition `m_pluginData != nullptr`, but it looks like it's guaranteed to continue on to the end of `NotePlayHandle::play(...)` where `m_totalFramesPlayed` is then increased (assuming no exceptions are thrown). This all means that `NotePlayHandle::play(...)` cannot set `m_totalFramesPlayed` to zero, always increases it before `m_instrumentTrack->playNote(...)` is called again, and therefore cannot create the condition where `m_totalFramesPlayed == 0 && m_pluginData != nullptr`. So possibility (2) can also be ignored.

This leaves the last possibility that I can see, which is (3). If `NotePlayHandle::resize(...)` were to be called sometime after the first time `m_instrumentTrack->playNote(...)` is called (so that `m_pluginData != nullptr`), it could set `m_totalFramesPlayed` to zero from some non-zero value it previously was. This would happen when `m_origFrames == 0` or `m_origTempo == 0`. I assume `m_origTempo` is never 0, so I'll ignore it. `m_origFrames` is only set to 0 if NotePlayHandle is constructed with the `_frames` parameter ==0. `NotePlayHandle::resize(...)` is only called if `Song::setTempo()` is called, which tracks with what you said PhysSong.

All of this means that the situation where the condition `m_totalFramesPlayed == 0 && m_pluginData != nullptr` would occur - thus causing memory leaks in most instrument plugins - looks like this:
1) A NotePlayHandle is created with `_frames` == 0.
2) `NotePlayHandle::play(...)` is called, and then `m_instrumentTrack->playNote(...)` is called from that.
3) The song's tempo changes, which calls `NotePlayHandle::resize(...)` and sets `m_totalFramesPlayed` back to 0.
4) Step 2 happens again, causing a memory leak.

To test whether I could actually cause this to happen, I added the following assert to the Vibed `playNote` routine:

```c++
assert((n->m_pluginData == nullptr) == (n->totalFramesPlayed() == 0));
```

Then I looped a section of a track with lots of Vibed notes playing and heavy tempo automation. The assert soon failed, meaning that a memory leak would have occurred had it still been using the old `n->totalFramesPlayed() == 0 || n->m_pluginData == nullptr` or `n->totalFramesPlayed() == 0` condition like most other instrument plugins still currently use.

I propose that we fix those conditions in the other instrument plugins, and maybe add some asserts in the `NotePlayHandle` code to make it clear which conditions hold true, what assumptions are being made, help demystify it, etc. Does that sound okay @PhysSong?

[the remark this answers is not in the document]
Comment (messmerd) [reply plugins/Vibed/Vibed.cpp]: One question that remains is what the solution should look like for each `playNote` routine.
I'm thinking either this:

```c++
if (n->totalFramesPlayed() == 0 || n->m_pluginData == nullptr)
{
    if (n->m_pluginData == nullptr)
    {
        deleteNotePluginData(n);
    }
    
    // ...

    n->m_pluginData = /* new plugin data here */
}
```

Or this:

```c++
if (n->m_pluginData == nullptr)
{
    // ...
    n->m_pluginData = /* new plugin data here */
}
```
The first option basically resets the plugin data to the default state whenever `m_totalFramesPlayed` is set back to zero, which is how most plugins currently operate - intended or not - while still fixing the memory leak.

[the remark this answers is not in the document]
Comment (PhysSong) [reply plugins/Vibed/Vibed.cpp]: The latter sounds better to me.

**Comment by superpaik** [comment]
Works ok. Tested on Win10-64b

**Review by superpaik — APPROVED** [review]

**Comment by luzpaz** [comment]
Any updates on this?

**Comment by messmerd** [comment]
> Any updates on this?

I'd like to complete #6680 first, then fix the merge conflicts it causes with this branch.
But this branch is ready for testing/review - I just need people willing to do it.

**Comment by messmerd** [comment]
@Testers/reviewers

Here's a project file for testing #6669:
[bug6669.zip](https://github.com/LMMS/lmms/files/11265022/bug6669.zip)
It also contains screenshots showing that this PR fixes the bug.

And here are project files for testing #6671:
[bug6671.zip](https://github.com/LMMS/lmms/files/11265024/bug6671.zip)
This also contains screenshots showing that the bug has been fixed. Unlike #6669, the bug occurs when saving the project instead of loading it, so I've included two project files - one created and saved in master, and one created and saved in this dev branch.

NOTE: In both the master branch and this dev branch, LMMS occasionally crashes randomly during or after editing automation for the Vibed instrument. I haven't figured out how to reliably cause it yet, but I'll create a new issue for it once I look into it some more. Since it also occurs in master, it seems unrelated to anything contained in this PR and should not affect testing, but I just wanted to give a heads up in case anyone encounters it.

**Review by superpaik — APPROVED** [review]

File: plugins/Organic/Organic.cpp [hunk 2]
```diff
@@ -159,61 +155,58 @@ OrganicInstrument::~OrganicInstrument()
...
+	for (int i = 0; i < m_numOscillators; ++i)
 	{
-		QString is = QString::number( i );
-		m_osc[i]->m_volModel.loadSettings( _this, "vol" + is );
-		if( _this.hasAttribute( "detune" + is ) )
+		const auto is = QString::number(i);
+
+		m_osc[i]->m_volModel.loadSettings(elem, "vol" + is);
+
+		if (elem.hasAttribute("detune" + is))
```
[the remark this answers is not in the document]
Comment (messmerd) [reply plugins/Organic/Organic.cpp]: The "detune" key [hasn't been used since 2014](https://github.com/LMMS/lmms/commit/63747787e414b9511625dd6c61f028886bb213c7), but it should be fixed with the changes I just pushed.

**Review by PhysSong — APPROVED** [review]

File: plugins/Organic/Organic.cpp [hunk 3]
```diff
@@ -159,61 +155,59 @@ OrganicInstrument::~OrganicInstrument()
...
-		QString is = QString::number( i );
-		m_osc[i]->m_volModel.saveSettings( _doc, _this, "vol" + is );
-		m_osc[i]->m_panModel.saveSettings( _doc, _this, "pan" + is );
-		m_osc[i]->m_harmModel.saveSettings( _doc, _this, "newharmonic" + is );
-
-		m_osc[i]->m_detuneModel.saveSettings( _doc, _this, "newdetune"
-									+ is );
-		m_osc[i]->m_oscModel.saveSettings( _doc, _this, "wavetype"
-									+ is );
+		QString is = QString::number(i);
```
Comment (Veratil) [inline plugins/Organic/Organic.cpp:166]: Change to be consistent with code elsewhere.
```suggestion
		const auto is = QString::number(i);
```

File: plugins/Vibed/Vibed.cpp [hunk 4]
```diff
@@ -52,628 +51,472 @@ extern "C"
...
+	connect(m_stringSelector.get(), SIGNAL(NineButtonSelection(int)), this, SLOT(showString(int)));
+
+	showString(0);
+
+	m_smoothBtn.move(79, 129);
+	m_smoothBtn.setActiveGraphic(PLUGIN_NAME::getIconPixmap("smooth_active"));
+	m_smoothBtn.setInactiveGraphic(PLUGIN_NAME::getIconPixmap("smooth_inactive"));
+	m_smoothBtn.setChecked(false);
+	m_smoothBtn.setToolTip(tr("Smooth waveform"));
+	connect(&m_smoothBtn, SIGNAL(clicked()), this, SLOT(smoothClicked()));
```
Comment (Veratil) [inline plugins/Vibed/Vibed.cpp:398]: Just a note for later enhancements: I feel we could change this to a function and just pass the variable and parameters to it.

File: plugins/Vibed/VibratingString.h [hunk 5]
```diff
@@ -34,244 +36,219 @@ namespace lmms
...
 			{
-				r = static_cast<float>( rand() ) /
-						RAND_MAX;
-				offset =  ( m_randomize / 2.0f -
-						m_randomize ) * r;
-				_dl->data[i] = _scale *
-						_values[_dl->length - i - 1] +
-						offset;
+				r = static_cast<float>(std::rand()) / RAND_MAX;
+				offset =  (m_randomize / 2.0f - m_randomize) * r;
```
Comment (Veratil) [inline plugins/Vibed/VibratingString.h:116]:
```suggestion
				offset = (m_randomize / 2.0f - m_randomize) * r;
```

**Review by Veratil — CHANGES_REQUESTED** [review]
Another small thing, we can probably remove the variable names in the declarations of class functions in the header files.

**Comment by luzpaz** [comment]
Soft bump

File: plugins/FreeBoy/FreeBoy.cpp [hunk 6]
```diff
@@ -250,7 +250,7 @@ void FreeBoyInstrument::playNote(NotePlayHandle* nph, sampleFrame* workingBuffer
 	int data = 0;
 	int freq = nph->frequency();
 
-	if ( tfp == 0 )
+	if (!nph->m_pluginData)
```
[the remark this answers is not in the document]
Comment (PhysSong) [reply plugins/FreeBoy/FreeBoy.cpp:253]: This is the correct form since tempo automation may cause creating note data more than once.

[the remark this answers is not in the document]
Comment (messmerd) [reply plugins/FreeBoy/FreeBoy.cpp:253]: PhysSong is right. Here's the investigation I did into this issue: https://github.com/LMMS/lmms/pull/6670#discussion_r1141261115

**Review by Veratil — APPROVED** [review]

**Comment by luzpaz** [comment]
Sweet!
